core/include/msg_bus.h

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     core_msg
+ *
+ * @{
+ *
+ * @file
+ * @brief       Messaging Bus API for inter process message broadcast.
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#ifndef MSG_BUS_H
+#define MSG_BUS_H
+
+#include <stdint.h>
+
+#include "list.h"
+#include "thread.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief A message bus is just a list of subscribers.
+ */
+typedef list_node_t msb_bus_t;
+
+/**
+ * @brief Static initializer for @ref msb_bus_t.
+ */
+#define MSG_BUS_INIT    {0}
+
+/**
+ * @brief Message bus subscriber entry.
+ *        Should not be modified by the user.
+ */
+typedef struct {
+    msb_bus_t next;         /**< next subscriber */
+    uint32_t event_mask;    /**< Bitmask of event classes */
+    kernel_pid_t pid;       /**< Subscriber PID */
+} msg_bus_entry_t;
+
+/**
+ * @brief Attach a thread to a message bus.
+ *
+ * This attaches a message bus subscriber entry to a message bus.
+ * Subscribe to events on the bus using @ref msg_bus_detach.
+ * The thread will then receive events with a matching type that are
+ * posted on the bus.
+ *
+ * Events can be received with @ref msg_receive.
+ *
+ * @param[in] bus           The message bus to attach to
+ * @param[in] entry         Message bus subscriber entry
+ */
+static inline void msg_bus_attach(msb_bus_t *bus, msg_bus_entry_t *entry)
+{
+    entry->next.next = NULL;
+    entry->event_mask = 0;
+    entry->pid = sched_active_pid;
+
+    list_add(bus, &entry->next);
```

Comment:
If you use clist, this would be O(1) instead of having to traverse the list. Also, clist code is shared by the scheduler and mutex/msg, so would probably reduce codesize.

Reply:
attach / detach are usually only done once per thread, so there is not much so save here.

I've changed `msg_bus_detach` to just be a wrapper around `list_remove`, if the thread somehow looses context of it's `msg_bus_entry_t`, I added `msg_bus_get_entry` for that.

Reply:
> attach / detach are usually only done once per thread, so there is not much so save here.

should it be guarded by irq_disable()/restore()?

Reply:
> I've changed `msg_bus_detach` to just be a wrapper around `list_remove`, if the thread somehow looses context of it's `msg_bus_entry_t`, I added `msg_bus_get_entry` for that.

:+1: